io.greenfire.Greenery v0.9.8 update, Wayland Error features=UseOzoneP…

tmrdev opened the pull request:
App update to v0.9.8, adding new options for Wayland socket and x11 fallback.

Also came across a new issue with running Greenery from a Flathub local build.

flatpak run io.greenfire.Greenery 

[2:0403/170448.706406:ERROR:ozone_platform_x11.cc(247)] Missing X server or $DISPLAY
[2:0403/170448.706429:ERROR:env.cc(226)] The platform failed to initialize.  Exiting.

I can run the app from the binary build (outside of Flathub runtime) with:

$: ./greenery --ozone-platform-hint=auto

Found this issue that details more:

https://github.com/element-hq/element-desktop/issues/763

Need to find the best way to provide a fix. In addition using the ozone auto option does not allow me to resize the Greenery App Window.

Also I left socket=x11 in the json config file.  This was showing Legacy Windowing System in App Details but need to see if fallback option will remove this. Otherwise remove the x11 option.

tmrdev commented:
> App update to v0.9.8, adding new options for Wayland socket and x11 fallback.
> 
> Also came across a new issue with running Greenery from a Flathub local build.
> 
> flatpak run io.greenfire.Greenery
> 
> [2:0403/170448.706406:ERROR:ozone_platform_x11.cc(247)] Missing X server or $DISPLAY [2:0403/170448.706429:ERROR:env.cc(226)] The platform failed to initialize. Exiting.
> 
> I can run the app from the binary build (outside of Flathub runtime) with:
> 
> $: ./greenery --ozone-platform-hint=auto
> 

I am also having issues getting Keybase (chat app that is based on Electron) to open.  I installed Keybase directly from the developers website so it was not installed from Flathub so this looks like a system/user config issue with Electron and Wayland.  I am also using Visual Studio (Electron base)  which was installed from Flathub and it is working ok.

razzeee commented:
It still works fine for me with that build, if I go back to only x11 enabled (running on wayland)

I think to make electron with wayland work, you would need to do something similar to https://github.com/flathub/im.riot.Riot/blob/master/element.sh and detect it on startup

tmrdev commented:
> It still works fine for me with that build, if I go back to only x11 enabled (running on wayland)
> 
> I think to make electron with wayland work, you would need to do something similar to https://github.com/flathub/im.riot.Riot/blob/master/element.sh and detect it on startup

Thank you for leading me to the bash script example, I will take a look.

I believe I've traced the issue to the $DISPLAY env.  Something may be modifying it and I'm debugging the process. On top of that I installed Tiger VNC and it was very buggy.  I disabled the vnc service so it shouldn't be modifying DISPLAY in anyway but I might clean it out entirely.  I may be able to get things working without using the bash script.

tmrdev commented:
> It still works fine for me with that build, if I go back to only x11 enabled (running on wayland)

When you go back to only x11 running wayland what does that mean?  How are you achieving this result?

Are you referencing when you login to the Windowing system and select which one you want to use like Gnome Class or Gnome Wayland.  With my testing on Fedora 39, Wayland is the default option and listed only as Gnome.  This is what I'm using currently to test the issue.

> I think to make electron with wayland work, you would need to do something similar to https://github.com/flathub/im.riot.Riot/blob/master/element.sh and detect it on startup

I went through all of the Element app Manifest and noticed the following:

- In the build-commands for Element there is this:

  - patch-desktop-filename ${FLATPAK_DEST}/Element/resources/app.asar
            
 Is it needed for Wayland support or not related?

- App Name is in capitals and they are setting special permissions, is this needed for Greenery
             
 - mkdir /app/Element 
 - cp -r * /app/Element
 - chmod -R a-s,go+rX,go-w "/app/Element"

- In the desktop file they are adding %U, is this needed?

 - Exec=/app/bin/element %U

I also saw this command referenced in the metainfo.xl description

flatpak run im.riot.Riot --enable-features=UseOzonePlatform,WebRTCPipeWireCapturer --ozone-platform=wayland

I can run this command for Greenery and the App will launch but I am unable to Resize the App Window.  

Also I'm going to remove any references to WebRTCPipeWireCapturer or similar as this should be related to the Chat functionality in the Element app?

I still need to implement the updated Shell script so I will see if that fixes it.  I have tested that adding --enable-features=WaylandWindowDecorations to the flatpak run command brings in the App Window Bar

tmrdev commented:
I have added Wayland support based on how the Element app does it.  It did not include any WebRTC references.  

Greenery launches and the App Window Bar works well along with the app.

I have also removed the x11 option from the finish args as the lint checker was complaining about it.

tmrdev commented:
> Build [113486 successful](https://buildbot.flathub.org/#/builders/6/builds/113486) To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/96294/io.greenfire.Greenery.flatpakref
> ```

I tested this build and everything looks good.  Should I add Wayland Support to the release notes in the description?

razzeee commented:
> When you go back to only x11 running wayland what does that mean? How are you achieving this result?

> Are you referencing when you login to the Windowing system and select which one you want to use like Gnome Class or Gnome Wayland. With my testing on Fedora 39, Wayland is the default option and listed only as Gnome. This is what I'm using currently to test the issue.

No, just removing the wayland permission, but still using a wayland session, so that it will get rendered by XWayland

tmrdev commented:
> > When you go back to only x11 running wayland what does that mean? How are you achieving this result?
> 
> > Are you referencing when you login to the Windowing system and select which one you want to use like Gnome Class or Gnome Wayland. With my testing on Fedora 39, Wayland is the default option and listed only as Gnome. This is what I'm using currently to test the issue.
> 
> No, just removing the wayland permission, but still using a wayland session, so that it will get rendered by XWayland

Xwayland is used for when there are issues with running an app with Wayland, isn't XWayland basically X11?

It looks like I can enable XWayland mode in Fedora 39 with this:

------------

If you are running a Wayland session, you can switch to XWayland mode by setting the WaylandEnable=false option in the /etc/gdm/custom.conf file.[2](https://www.reddit.com/r/Fedora/comments/18fpt4e/okay_how_to_switch_to_wayland/)[3](https://unix.stackexchange.com/questions/382542/how-to-use-wayland-in-fedora)

Specifically, you would edit the /etc/gdm/custom.conf file and change the line: WaylandEnable=true to WaylandEnable=false This will force your Wayland session to use the XWayland compatibility layer to run X11 applications.[4](https://docs.fedoraproject.org/en-US/fedora/latest/system-administrators-guide/Wayland/)


------------

I will check my system (it should already be configured), in theory a Wayland app should fallback to X11 if it is unable to run natively in Wayland.

tmrdev commented:
I'd like to Merge this PR so v0.9.8 for Greenery can be pushed.  Should I do this myself or does a Flathub maintainer have to do it?

razzeee commented:
You can do it if you're happy, flathub maintainers mostly only get active with apps, that have been abandoned.
